### PR TITLE
Smarter row build state

### PR DESCRIPTION
**Pull request by neworderofjamie**
This P.R. slightly extends the syntax for defining connectivity initialization models so you can initialize row-wide state variables with expressions (which get substitutions applied etc) as well as with constants. This lets you express the fixed total number of synapses connectivity in such a way that the row length is only read once per row. This actually makes a pretty sizable difference to the performance - 20% faster procedural synapse update on my home machine. Because it uses the same code path, this functionality also now applies to neuron additional inputs although I can't think of a use-case off hand...

**NOTE** this is currently set to merge into #293 as it updates the built in models

### CODE REVIEW

**Comment by neworderofjamie** [comment]
Sadly that 20% doesn't translate to the multi-area simulation - presumably as, for once, there is enough computation occuring to hide memory latency!

File: include/genn/genn/code_generator/codeGenUtils.h [hunk 1]
```diff
@@ -106,44 +101,6 @@ bool regexFuncSubstitute(std::string &s, const std::string &trg, const std::stri
 void functionSubstitute(std::string &code, const std::string &funcName,
                         unsigned int numParams, const std::string &replaceFuncTemplate);
 
-//--------------------------------------------------------------------------
-//! \brief This function writes a floating point value to a stream -setting the precision so no digits are lost
-//--------------------------------------------------------------------------
-template<class T, typename std::enable_if<std::is_floating_point<T>::value>::type* = nullptr>
```
Comment (tnowotny) [inline include/genn/genn/code_generator/codeGenUtils.h]: I am not sure I can parse this ... is it the same as
```std::enable_if_t<std::is_floating_point<T>::value, int> = 0``` ?
Oh boy - feels like I have gotten totally out of touch with more recent C++ standards.

**Comment by neworderofjamie** [comment]
Yeah, the enable_if syntax is particularly awful - I seem to remember this article helped me grasp it https://eli.thegreenplace.net/2014/sfinae-and-enable_if/. All that's happening with that function is that I'm moving it to a different utils module though! Give me a shout when you're around tomorrow and I can attempt to explain though!

**Comment by tnowotny** [comment]
Thanks for the link. I (almost) get it now. Yes, I realised that this had nothing to do with the pull request. I didn't really find the meat of the pull request in the file changes, though.

**Comment by neworderofjamie** [comment]
Hurrah! What 'meat' there is as follows:

1. Switch the ``ParamVal`` structure, used for specifying these state variables, to use a string rather than a double as it's value type and add a constructor taking a double for backward compatibility - https://github.com/genn-team/genn/blob/smarter_row_build_state/include/genn/genn/snippet.h#L138-L143
2. Apply suitable substitutions to the string when writing it out in synaptic matrix initialisation code (https://github.com/genn-team/genn/blob/smarter_row_build_state/src/genn/genn/code_generator/generateInit.cc#L344-L348) and procedural connectivity code (https://github.com/genn-team/genn/blob/smarter_row_build_state/src/genn/genn/code_generator/generateSynapseUpdate.cc#L161-L165)
2. Use the new features in some sparse connectivity init snippets - https://github.com/genn-team/genn/blob/smarter_row_build_state/include/genn/genn/initSparseConnectivitySnippet.h#L277-L287

**Pull request merged**